src/main/java/com/microsoft/graph/concurrency/ChunkedUploadResponseHandler.java

```diff
@@ -98,35 +100,38 @@ public void configConnection(@Nonnull final Response response) {
 		if (response.code() >= HttpResponseCode.HTTP_CLIENT_ERROR) {
 			logger.logDebug("Receiving error during upload, see detail on result error");
 
-			return new ChunkedUploadResult<UploadType>(
+			return new ChunkedUploadResult<>(
 					GraphServiceException.createFromResponse(request, null, serializer,
-						response, logger));
-		} else if (response.code() >= HttpResponseCode.HTTP_OK
+							response, logger));
+		} else if (response.code() >= HTTP_OK
 				&& response.code() < HttpResponseCode.HTTP_MULTIPLE_CHOICES) {
 			final Map<String, String> headers = responseHeadersHelper.getResponseHeadersAsMapStringString(response);
 			final String contentType = headers.get(Constants.CONTENT_TYPE_HEADER_NAME);
 			final String location = headers.get("Location");
-			if(contentType != null
-				&& contentType.contains(Constants.JSON_CONTENT_TYPE)) {
+			if (contentType != null
+					&& contentType.contains(Constants.JSON_CONTENT_TYPE)) {
 				try (final InputStream in = new BufferedInputStream(response.body().byteStream())) {
-					final String rawJson = CoreHttpProvider.streamToString(in);
-					final IUploadSession session = serializer.deserializeObject(rawJson, uploadSessionClass);
-					if(session == null || session.getNextExpectedRanges() == null) {
+					if (isUploadSessionCompleted(response)) {
 						logger.logDebug("Upload session is completed (ODSP), uploaded item returned.");
-						final UploadType uploadedItem = serializer.deserializeObject(rawJson, this.deserializeTypeClass);
-						return new ChunkedUploadResult<UploadType>(uploadedItem);
+						final UploadType uploadedItem = serializer.deserializeObject(in, this.deserializeTypeClass);
+						return new ChunkedUploadResult<>(uploadedItem);
 					} else {
+						final IUploadSession session = serializer.deserializeObject(in, uploadSessionClass);
 						logger.logDebug("Chunk bytes has been accepted by the server.");
-						return new ChunkedUploadResult<UploadType>(session);
+						return new ChunkedUploadResult<>(session);
 					}
 				}
-			} else if(location != null) {
+			} else if (location != null) {
 				logger.logDebug("Upload session is completed (Outlook), uploaded item returned.");
-				return new ChunkedUploadResult<UploadType>(this.deserializeTypeClass.getDeclaredConstructor().newInstance());
+				return new ChunkedUploadResult<>(this.deserializeTypeClass.getDeclaredConstructor().newInstance());
 			} else {
 				logger.logDebug("Upload session returned an unexpected response");
 			}
 		}
-		return new ChunkedUploadResult<UploadType>(new ClientException("Received an unexpected response from the service, response code: " + response.code(), null));
+		return new ChunkedUploadResult<>(new ClientException("Received an unexpected response from the service, response code: " + response.code(), null));
+	}
+
+	private boolean isUploadSessionCompleted(Response response) {
```

Comment:
I understand the performance benefit of avoiding tryin deserializing an upload session when it might not be one, but it was this way on purpose: the services (OneDrive, OneDrive for Business, Outlook, Exchange) behave differently from one another and the only alignment is in the upload session response. Can you revert that part of the change please?

Reply:
resolving as it's been double checked and the differences have been handled in subsequent comments